examples/widgets/search_entry.nim

```diff
@@ -0,0 +1,58 @@
+# MIT License
+# 
+# Copyright (c) 2023 Can Joshua Lehmann
+# 
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+# 
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+# 
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE
+
+import owlkettle, owlkettle/[playground, adw, dataentries]
+
+viewable App:
+  searchDelay: uint = 100
+  placeholderText: string = "Search"
+  sensitive: bool = true
+  tooltip: string = ""
+  sizeRequest: tuple[x, y: int] = (-1, -1)
+  
+method view(app: AppState): Widget =
+  result = gui:
+    Window():
+      defaultSize = (400, 120)
+      title = "Search Entry Example"
+      HeaderBar() {.addTitlebar.}:
+        insert(app.toAutoFormMenu()) {.addRight.}
+      
+      Box(orient = OrientY):
+        Label(text = "Search something") {.expand: false.}
+        SearchEntry() {.expand: false.}:
+          searchDelay = app.searchDelay
+          placeholderText = app.placeholderText
+          sensitive = app.sensitive
+          tooltip = app.tooltip
+          sizeRequest = app.sizeRequest
+          
+          proc activate(searchString: string) =
+            echo "Search triggered: ", searchString
+          
+          proc nextMatch() = echo "Next Match"
+          proc previousMatch() = echo "Prior Match"
+          proc searchStarted(searchString: string) = echo "Search Started: ", searchString
+          proc changed(searchString: string) = echo "Search Changed: ", searchString
+          proc stopSearch(searchString: string) = echo "Search Stopped: ", searchString
```

Comment:
I would put an actual searchable list in this example (just a `ListBox` in a `ScrollableWindow` with ~100 autogenerated items), so that it demonstrates what the nextMatch, previousMatch, ... callbacks should do exactly. Also not sure what the HIG says about placement of search entries, but I am under the impression that putting it in the Header bar might be nicer in that case.

Reply:
I did tackle this, not sure if that is to your satisfaction though.
Introduced a couple more fields to the example (selected, items and filteredItems), triggered filtering of the items as you type in the searchbar, I only render the filtered items and mark the one that is selected as bold.
With the nextMatch/previousMatch you can then shift the index around and by hitting enter it prints out what you currently have selected.

Reply:
Fixed the spacing issue, I was silly and forgot an expand: false for the label currently being the "target"